src/potdata/samplers.py

```diff
@@ -616,3 +616,244 @@ def plot2(self, separation1: int, separation2: int, show: bool = False,
 
         if show:
             plt.show()
+
+class KMeansStructureSampler(BaseStructureSampler):
+    """Sample structures using KMeans clustering.
+
+    Args:
+        kmeans_kwargs: Arguments to pass to `sklearn.cluster.KMeans`. You should
+            provide the `n_clusters` argument to specify the number of clusters.
+    """
+
+    DEFAULT_SOAP_KWARGS = {"r_cut": 5.0, "n_max": 8, "l_max": 5, "periodic": True}
+    DEFAULT_KMEANS_KWARGS = {"n_clusters": 8}
+
+    def __init__(
+        self,
+        soap_kwargs: dict = None,
+        species_to_select: list[str] | None = None,
+        pool_method: str = "concatenate",
+        pca_dim: int | None = None,
+        kmeans_kwargs: dict = None,
+        ratio: float = 1.0,
+        seed: int = 35,
+    ):
+        self.soap_kwargs = self.DEFAULT_SOAP_KWARGS.copy()
+        if soap_kwargs is not None:
+            self.soap_kwargs.update(soap_kwargs)
+
+        self.kmeans_kwargs = self.DEFAULT_KMEANS_KWARGS.copy()
+        if kmeans_kwargs is not None:
+            self.kmeans_kwargs.update(kmeans_kwargs)
+
+        self.species_to_select = species_to_select
+        self.pool_method = pool_method
+        self.pca_dim = pca_dim
+        self.ratio = ratio
```

Comment:
How about checking the range of `ratio` if it is a float? change line 652 to: 
```python
if isinstance(ratio, float) and not 0.0 < ratio < 1.0:
    raise Value(f"Expect ratio to be an integer or a float between 0.0 and 1.0. Got {ratio}.")
self.ratio = ratio
```

Reply:
Done, thanks